### PR TITLE
[spec] v2.2 - action chaining

### PR DESCRIPTION
**Pull request by nickfrosty**
## TLDR

Actions can be chained together in successive series up to a depth of `n`. Each action can provide a callback url to fetch the next action in the series (being provided the `signature` from the previous action`) which can then be presented to users via action aware clients.

Action chaining allows developer to build more complex and dynamic experiences within blinks, including:

- providing multiple transactions (and eventually sign message) to a user
- customized action metadata based on the user's wallet address
- refreshing the blink metadata after a successful transaction
- receive an API callback with the transaction signature for additional
  validation and logic on the Action API server

## Rationale

Blinks and actions are currently limited to a single depth of interaction. After the user signs a transaction, it concludes the experience. Action chaining allows developers to provide a depth of `n` actions, creating more dynamic and comprehensive experiences.

sRFC: https://forum.solana.com/t/srfc-28-blinks-chaining/1734

### CODE REVIEW

File: packages/actions-spec/README.md [hunk 1]
```diff
@@ -498,6 +532,75 @@ must do so only if a signature for the `account` in the request is expected.
...
+Action's transaction is confirmed on-chain, the next action can be obtained and
+presented to the user.
+
+Action chaining allows developer to build more complex and dynamic experiences
+within blinks, including:
+
+- providing multiple transactions (and eventually sign message) to a user
+- customized action metadata based on the user's wallet address
+- refreshing the blink metadata after a successful transaction
+- receive an API callback with the transaction signature for additional
```
Comment (tsmbl) [inline packages/actions-spec/README.md:547]: Also proposing to explicitly highlight the behaviour on tx failed during confirmation - my proposal is to always send the tx signature to API regardless of confirmation status to enable basic error handling in all cases
* tx confirmed
* tx failed
* tx timed out

Thoughts?

File: packages/actions-spec/README.md [hunk 2]
```diff
@@ -441,11 +459,21 @@ A `POST` response with an HTTP `OK` JSON response should include a body payload
...
   /** base64 encoded serialized transaction */
   transaction: string;
   /** describes the nature of the transaction */
   message?: string;
+  links?: {
+    /**
+     * the next action in a successive chain of actions to be obtained and/or rendered
+     * to the user after the previous was successful
+     */
+    next: string | NextAction<T | "action">;
```
[the remark this answers is not in the document]
Comment (tsmbl) [reply packages/actions-spec/README.md:475]: May be the `embedded` is not the best noun, but the idea should be clear

File: packages/actions-spec/README.md [hunk 3]
```diff
@@ -160,7 +160,16 @@ A `GET` response with an HTTP `OK` JSON response should include a body payload
...
+export type ActionType = "action" | "completed";
+
+export type ActionGetResponse = Action<"action">;
+
+/**
+ * A single Solana Action
+ */
+export interface Action<T extends ActionType = "action"> {
+  /** @default `action` */
+  type?: T;
```
[the remark this answers is not in the document]
Comment (nickfrosty) [reply packages/actions-spec/README.md:172]: Sure, it accomplishes the same goal. So I'm fine with this

File: packages/actions-spec/README.md (hunk 1 above)
Comment (nickfrosty) [reply packages/actions-spec/README.md:547]: If the transaction failed to confirm, I would assume the action could be called again via the blink UI to attempt to repeat it. Is this not the case?

Having the ability to achieve a callback based on each of these solana transaction lifecycle events could be useful though. But it does feel like scope creep...

- How do you propose the blink-client denotes each of these lifecycle statuses?
- What should the UI do after the getting a response from the callback?
- Should only a single callback url for all these lifecycle events be supported? Or should the developer be able to define specific callback urls for each lifecycle event?

File: packages/actions-spec/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (nickfrosty) [reply packages/actions-spec/README.md:475]: We can avoid the string easily with the suggestion you made in [this comment](https://github.com/solana-developers/solana-actions/pull/18#discussion_r1697431298) :)

To me, making this union types accomplishes the same thing but with more complexity in the type and makes a developer have to write more verbose code to accomplish the same thing. Especially since we are only supporting two types of "next actions" right now.

If and when more types are supported, we might then need the more complicated type and verbosity you are suggesting here. It would be easily compatible with the current proposal AND your suggestion here by performing a union type based on the `type` field then.

File: packages/actions-spec/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickfrosty) [reply packages/actions-spec/README.md:172]: I just pushed this update to the PR

File: packages/actions-spec/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsmbl) [reply packages/actions-spec/README.md:475]: I think the comment you are referring in comment is actually related, but expresses the same concept, I'm proposing here. The idea of the comment was to enforce `type` attribute being mandatory in `NextAction`, since it was optional. This way action developers will be enforced to set the `type` and client can then deterministically distinguish if the next action is `completed` or `action`.

While the current approach accomplishes the task, we should also consider extensibility, type-safety and deterministic interpretation of data structures in the future.

### Main Problems with raw strings and primitive values in general

The primary issue with using raw strings and primitive values in general is extensibility. It's generally best practice to wrap plain values in objects to allow for additional attributes or have other union options that may be expressed with the same primitive type, but have different semantics.

####  Example scenarios illustrating the potential problems:

1. **Supporting GET URLs in the future:**
   - **Current approach:**
     ```typescript
     export interface ActionPostResponse<T extends ActionType = ActionType> {
       transaction: string;
       message?: string;
       links?: {
         next: string | NextAction<T | "action">;
       };
     }
     ```
   - If we decide to support GET URLs in the future, a plain string won't suffice, and we will need to add an object.
     ```typescript
     export interface GetNextActionLink {
       type: "get";
       href: string;
     }
     ```
   - This would require maintaining the raw string for backward compatibility and introducing a new object, leading to an inconsistent interface - POST will be raw string, GET will be an object.

2. **Adding extra parameters to POST:**
   - **Current Approach:**
     ```typescript
     export interface ActionPostResponse<T extends ActionType = ActionType> {
       transaction: string;
       message?: string;
       links?: {
         next: string | NextAction<T | "action">;
       };
     }
     ```
   - If we want to add extra parameters to a POST action, using a raw string will make this impossible. We would need to introduce an object to handle additional parameters.
     ```typescript
     export interface PostNextActionLink {
       type: "post";
       href: string;
       extraParams?: { [key: string]: any };
     }
     ```
   - Again, this would require keeping the raw string for backward compatibility and introducing a new object, resulting in inconsistency.

In both cases, maintaining backward compatibility means supporting the raw string while introducing new objects as proposed. This will complicate the interface and make it look inconsistent, decreasing the DX.

### Need for type attribute in union type members

When you have multiple objects, a clear way to distinguish the type client-side is necessary for type-safe switch-casing. Using `typeof` won't work effectively. Instead, a `type` attribute ensures deterministic interpretation and easy handling.

#### Example:

**Proposed approach:**
```typescript
export type NextActionLink<T extends ActionType> = PostNextActionLink | GetNextActionLink | EmbeddedNextAction<T>;

export interface PostNextActionLink {
  type: "post";
  href: string;
}

export interface GetNextActionLink {
  type: "get";
  href: string;
}

export interface EmbeddedNextAction<T extends ActionType> {
  type: "embedded";
  action: NextAction<T>;
}

function handleNextAction(next: NextActionLink<any>) {
  switch (next.type) {
    case 'post':
      // Type safe handle post action with next.href
      break;
    case 'get':
      // Type safe handle get action with next.href
      break;
    case 'embedded':
      // Type safe handle embedded action with next.action
      break;
    default:
      // Handle unexpected case
      break;
  }
}

```

### References to data structures designed for similar problem space
Many well-designed APIs follow these best practices:

- **HAL:** Check how `next` is defined in HAL:
  [HAL Specification](https://stateless.group/hal_specification.html)
- **Farcaster Actions:** Look at how actions are defined in Farcaster, even for a single action type:
  [Farcaster Actions Specification](https://docs.farcaster.xyz/reference/actions/spec#properties)

These references are designed with the same reasons highlighted above: extensibility, consistency, and deterministic client-side interpretation during the future updates.

File: packages/actions-spec/README.md (hunk 1 above)
Comment (tsmbl) [reply packages/actions-spec/README.md:547]: I think we generally have 2 options

a) As you proposing, assume the action should be called again via the blink UI to attempt to repeat it. I think it will work fine, but can be a bit annoying if you have a long chain and want to repeat only last action.
b) If next action is defined, always pass the tx signature to backend regardless of confirmation result - backend can check signature status and make a decision about the next step, e.g. retrying last action, rather than to repeat the entire chain from the beginning

So, blink client in (b) just needs to always send tx signature to backend, backend then will return the next state that should be shown

> How do you propose the blink-client denotes each of these lifecycle statuses?
No need for special behaviour, just always pass tx signature after tx confirmation stage finished
> What should the UI do after the getting a response from the callback
Just render it, as it would do it for the any other next action
> Should only a single callback url for all these lifecycle events be supported? Or should the developer be able to define specific callback urls for each lifecycle event?
I think single would be already good for the start - developers can just check signature status on their backend to decide what should be done next

File: packages/actions-spec/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (nickfrosty) [reply packages/actions-spec/README.md:475]: I totally get understand your goal and desired implementation, I just think it is more complicated and verbose. Both of our recommendations accomplish the same thing and effectively provide the same type safety and DX (for both action developer and blink clients). 

As a dev building actions, I would prefer using my recommended implementation but I can also see why as a dev building blink clients, you would prefer yours. The accomplish the same goal in the end.

I'm fine with moving forward with yours so we can finalize the spec and get it shipped. Can you just suggest the changes on my PR?

File: packages/actions-spec/README.md (hunk 1 above)
Comment (nickfrosty) [reply packages/actions-spec/README.md:547]: >a) As you proposing, assume the action should be called again via the blink UI to attempt to repeat it. I think it will work fine, but can be a bit annoying if you have a long chain and want to repeat only last action.

I would not expect the blink-client to attempt to repeat the entire action chain from the start if the current action failed to confirm. That would be foolish lol. I would only the client to reattempt the current action.

> No need for any special behaviour, just always pass tx signature to callback after tx confirmation stage finished

there would have to be some way to denote each of the lifecycle status you suggested ones though: `tx confirmed`, `tx failed`, and `tx timed out`. 

A transaction can fail for many reason on the client side. "tx failed" would be a generic failure vs "tx timeout" would be a specific one like an expired blockhash. it could fail preflight checks performed by the blink-client and or wallet. 

many causes of a transaction failing of which would result in no tx id existing on chain. Therefore not giving much useful info to the action api to actually handle various error states. there would not always be a tx id to give to process on-chain errored transaction

> Does it still feels like scope creep to you?
It does, but if we opt for a simple design that we can flush out quickly, I'm okay with it being in this spec update. It will for sure be useful for devlopers!

Comment (tsmbl) [reply packages/actions-spec/README.md:547]: Long story short: I believe it's good enough to go with just reattempting the last action, we can improve error handling later as a separate sRFC.

> I would not expect the blink-client to attempt to repeat the entire action chain from the start if the current action failed to confirm. That would be foolish lol. I would only the client to reattempt the current action.

Ahh, ok, then I've misunderstood. Reattempting the last action instead of the entire chain is good for the start

> there would have to be some way to denote each of the lifecycle status you suggested ones though: tx confirmed, tx failed, and tx timed out.

In the future could still be a single callback, but with different payloads. Generally errors should have an error code + error message, that is useful info to pass to API together with the signature if it's available. E.g. see phantom deeplink callback https://docs.phantom.app/phantom-deeplinks/provider-methods/signandsendtransaction#reject. Not advocating to use the same structures, but this just gives an idea of data we could include.

File: packages/actions-spec/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (nickfrosty) [reply packages/actions-spec/README.md:475]: Merged your PR into mine :)

File: packages/actions-spec/README.md [hunk 4]
```diff
@@ -171,6 +177,7 @@ export interface ActionGetResponse {
   label: string;
   /** UI state for the button being rendered to the user */
   disabled?: boolean;
+  /**  */
```
Comment (fsher) [inline packages/actions-spec/README.md:180]: minor: missing description here

File: packages/actions-spec/README.md (hunk 1 above)
Comment (tsmbl) [reply packages/actions-spec/README.md:543]: To summarize, imo gtg with reattempting the last action

Let's skip error handling callback improvements and make them as a separate spec update, seems like we need broader discussion here

File: packages/actions-spec/README.md (hunk 4 above)
Comment (nickfrosty) [reply packages/actions-spec/README.md]: I just removed the empty comment :)

File: packages/actions-spec/README.md [hunk 5]
```diff
@@ -441,11 +455,21 @@ A `POST` response with an HTTP `OK` JSON response should include a body payload
...
   /** describes the nature of the transaction */
   message?: string;
+  links?: {
+    /**
+     * The next action in a successive chain of actions to be obtained after
+     * the previous was successful.
+     */
+    next: NextActionLink;
+  };
 }
```
Comment (thearyanag) [inline packages/actions-spec/README.md:473]: The `ActionPostResponse` should be as follows, allowing for form-like use-cases where tx approval is not required for each action and maybe just the last action in the chain required the approval.  


```
export interface ActionPostResponse<T extends ActionType = ActionType> {
    /**
     * If the current action requires a transaction to be signed then this should be a base64 encoded serialized transaction
     * If the current action does not require a transaction to be signed, then this should be null
     */
  transaction: string | null;
  /** describes the nature of the transaction */
  message?: string;
  links?: {
    /**
     * The next action in a successive chain of actions to be obtained after
     * the previous was successful.
     */
    next: NextActionLink;
  };
}
```

Comment (enzotar) [reply packages/actions-spec/README.md:473]: Form-like is needed. From a UX point of view, you don't want to sign more than once. So ideally, at every new screen, the developer gets new information and packages a final instruction for the user to sign. If you want to make users sign every screen, then you can too. 

I would even make every chain/screen transaction optional. Maybe you sign on the first screen and the last screen is an optional survey. Maybe there is no transaction at all, simply collecting info, e.g. pubkeys

Comment (thearyanag) [reply packages/actions-spec/README.md:473]: yes, that's why the transaction can be either `string` or `null`, making the transaction optional.

Comment (nickfrosty) [reply packages/actions-spec/README.md:473]: While this is a useful thing to support form-like experiences, actions are still currently required to return a transaction, since the user is required to sign a transaction.

Implementing something like this is planned in general, but not within this spec change / proposal. When sign message support is added, maybe in there?

@thearyanag maybe you could post a more complete sRFC with details about a proposal to support some sort of "non signing action" support? My gut is that something like this should not be allowed on the root action, so the user is still required to "initiate a session" within the blink by signing a transaction (or eventually a message), then they could complete form-like experiences

Comment (enzotar) [reply packages/actions-spec/README.md:473]: Why is the user required to sign transaction?

Comment (nickfrosty) [reply packages/actions-spec/README.md:473]: Because that is the core of what an action is right now. From the docs and spec:
> [Solana Actions](https://solana.com/docs/advanced/actions#actions) are specification-compliant APIs that return transactions on the Solana blockchain to be previewed, signed, and sent across a number of various contexts, including QR codes, buttons + widgets, and websites across the internet.

If an when this is to change, it would need to be a different sRFC and discussion. It would need additional discussion and thought, not to avoid scope creep on this individual PR / spec update.

Comment (enzotar) [reply packages/actions-spec/README.md:473]: The Blink will become inconvenient if you have to sign more than once. Unless it is a feature of that blink to sign multiple times, I believe requiring more than one signature is UX misstep.

Fill form, sign, fill new form, sign again. Devs stuck with partial signs to handle. Find the Blink again, load the partial properly.

At which point you might as well redirect the user to an external site to provide a streamline experience.

Comment (nickfrosty) [reply packages/actions-spec/README.md:473]: I agree with you on form based things. I agree that the idea of the user not being always required to sign something in every single action is a good one. It can unlock some interesting new experiences within blinks.

I'm saying it is out of scope for this specific PR/spec change. This spec change has already gone through much discussion on the Solana forum for several weeks now and is effectively finalized.

Hence my original request:
> post a more complete sRFC with details about a proposal to support some sort of "non signing action" support

Comment (nickfrosty) [reply packages/actions-spec/README.md:473]: Within the upcoming spec proposal for sign message, we are planning to suggest a session token that can be passed back and forth between the action api server and the blink client. Think JWT passed between server and client, but within a blink.

Something like this would also make the idea of non-signing actions even more powerful and useful. The user initiates an authenticated session (by signing a message via their wallet in the first action in a chain), the api server verifies the signature in their backend, the api server provides their `token` for the client interactions.

Off the top of my head:
Going with a pure "this is a form and does not require the user to sign anything" approach you seem to be suggesting is easily abused since a bot can spam the action api server to submit the form data. But never truly connect or verify a wallet. Ripe for abuse.

Hence my suggestion of:
> to be a different sRFC and discussion. It would need additional discussion and thought

Comment (enzotar) [reply packages/actions-spec/README.md:473]: Fair. I'm late to the discussion and this is a needed step anyhow.

Comment (thearyanag) [reply packages/actions-spec/README.md:473]: > While this is a useful thing to support form-like experiences, actions are still currently required to return a transaction, since the user is required to sign a transaction.
> 
> Implementing something like this is planned in general, but not within this spec change / proposal. When sign message support is added, maybe in there?
> 
> @thearyanag maybe you could post a more complete sRFC with details about a proposal to support some sort of "non signing action" support? My gut is that something like this should not be allowed on the root action, so the user is still required to "initiate a session" within the blink by signing a transaction (or eventually a message), then they could complete form-like experiences

agreed, I'll post a more detailed sRFC
